jwt/keyset.go

```diff
@@ -0,0 +1,217 @@
+// JSON Web Key Set (JWKS) published in the discovery document at the given discoveryURL.
+// The client used to obtain the remote keys will verify server certificates using the root
+// certificates provided by discoveryCAPEM.
+func NewOIDCDiscoveryKeySet(ctx context.Context, discoveryURL string, discoveryCAPEM string) (KeySet, error) {
+	if discoveryURL == "" {
+		return nil, errors.New("discoveryURL must not be empty")
+	}
+
+	caCtx, err := createCAContext(ctx, discoveryCAPEM)
+	if err != nil {
+		return nil, err
+	}
+
+	provider, err := oidc.NewProvider(caCtx, discoveryURL)
+	if err != nil {
+		return nil, err
+	}
+
+	return &OIDCDiscoveryKeySet{
+		provider: provider,
+	}, nil
+}
+
+// VerifySignature parses the given JWT, verifies its signature using discovered JWKS keys, and
+// returns the claims in its payload. The given JWT must be of the JWS compact serialization form.
+func (ks OIDCDiscoveryKeySet) VerifySignature(ctx context.Context, token string) (map[string]interface{}, error) {
+	// Verify only the signature
+	oidcConfig := &oidc.Config{
+		SkipClientIDCheck: true,
+		SkipExpiryCheck:   true,
+		SkipIssuerCheck:   true,
+	}
+
+	verifier := ks.provider.Verifier(oidcConfig)
+	idToken, err := verifier.Verify(ctx, token)
+	if err != nil {
+		return nil, err
+	}
+
+	allClaims := make(map[string]interface{})
+	if err := idToken.Claims(&allClaims); err != nil {
+		return nil, err
+	}
+
+	return allClaims, nil
+}
+
+// NewJSONWebKeySet returns a KeySet that verifies JWT signatures using keys from the JSON Web
+// Key Set (JWKS) at the given jwksURL. The client used to obtain the remote JWKS will verify
+// server certificates using the root certificates provided by jwksCAPEM.
+func NewJSONWebKeySet(ctx context.Context, jwksURL string, jwksCAPEM string) (KeySet, error) {
+	if jwksURL == "" {
+		return nil, errors.New("jwksURL must not be empty")
+	}
+
+	caCtx, err := createCAContext(ctx, jwksCAPEM)
+	if err != nil {
+		return nil, err
+	}
+
+	return JSONWebKeySet{
+		remoteJWKS: oidc.NewRemoteKeySet(caCtx, jwksURL),
+	}, nil
+}
+
+// VerifySignature parses the given JWT, verifies its signature using JWKS keys, and returns
+// the claims in its payload. The given JWT must be of the JWS compact serialization form.
+func (ks JSONWebKeySet) VerifySignature(ctx context.Context, token string) (map[string]interface{}, error) {
+	payload, err := ks.remoteJWKS.VerifySignature(ctx, token)
+	if err != nil {
+		return nil, err
+	}
+
+	// Unmarshal payload into a set of all received claims
+	allClaims := map[string]interface{}{}
+	if err := json.Unmarshal(payload, &allClaims); err != nil {
+		return nil, err
+	}
+
+	return allClaims, nil
+}
+
+// NewStaticKeySet returns a KeySet that verifies JWT signatures using PEM-encoded public keys.
+// The given publicKeys must be of PEM-encoded x509 certificate or PKIX public key forms.
+func NewStaticKeySet(publicKeys []string) (KeySet, error) {
```

Comment:
could we just accept []crypto.PublicKey directly and provide helper functions to parse PEMs and convert Certs to PublicKeys?  Is there an advantage to accepting PEMs?

Reply:
Will address along with comment on `StaticKeySet`.

Reply:
Addressed in d944a7e. Let me know your thoughts.